### PR TITLE
Backup config on flash

### PR DESCRIPTION
**Pull request by haslinghuis**
Saves the configuration on the flight controller before flashing.

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by McGiverGim** [comment]
> Error building NW apps: Error: certificate has expired

Some error with some certificate in the middle, I suppose on the nw.js server to download the nw zip. But it seems fixed now.

**Comment by limonspb** [comment]
I think the filename should also have date/time in the name? So we don't overwrite the old ones/previous drone flashes?

**Comment by TheIsotopes** [comment]
@haslinghuis Seems to be working so far.
`status`, `diff all` and `dump all` are saved to one file.

**Comment by nerdCopter** [comment]
this is fantastic!
please use the existing naming convention, just adding `_backup`.
e.g. BTFL_cli_backup_20230427_165129_SKYSTARSF405AIO.txt

File: src/js/utils/generate_filename.js
Comment (haslinghuis) [inline src/js/utils/generate_filename.js:19]: Perhaps we could add FC.CONFIG.boardName to this block and move it down to the line 41

**Comment by nerdCopter** [comment]
format good! 
if flashing over non-betaflight, backup file still includes `BTFL` -- can it use FC.CONFIG.flightControllerIdentifier instead?

**Comment by haslinghuis** [comment]
That is a minor problem found, FC.CONFIG.flightControllerIdentifier is not populated here.

Also should we allow craftName at all in generateFilename() ?

**Comment by haslinghuis** [comment]
- Fixed by calling MSP_FC_VARIANT (in detect board)
- Also cloud api will now only check if variant === "BTFL" (in detect board)

**Review by nerdCopter — APPROVED** [review]

**Comment by nerdCopter** [comment]
i jsut found a weird behavior. 
press enter inside of custom defines, and it launches the backup behavior.

**Comment by haslinghuis** [comment]
@nerdCopter found the enter key event:

image

**Comment by nerdCopter** [comment]
> @nerdCopter found the enter key event:
> 
> image

cool. maybe it's still wanted then? :man_shrugging:

**Comment by nerdCopter** [comment]
on second thought, i don't mind enter triggering flash, but it is **invalid** without loading firmware first. so maybe enter should trigger "load firmware" or maybe trigger both in proper order.
of course not trigger anything would be easiest fix.

**Comment by haslinghuis** [comment]
As the backup operation depends on auto-detect and is executed very close before flashing there is risk reconnecting to soon and will result in not being able to open the port.

Think we need to check some more scenario's:
- dfu mode
- no internet connection
- local flashing
- cancelling backup

User should be able to disable backup on flash in options (on by default) or simple open dialog where the user can click backup

**Review by nerdCopter — APPROVED** [review]

**Comment by nerdCopter** [comment]
i wonder if this should query the `full-chip erase` toggle as the deciding factor.

**Review by TheIsotopes — APPROVED** [review]

**Comment by haslinghuis** [comment]
Somehow the first flash is always robust, but the second time did not work for me at all. That's the weird part.
After some hours debugging and testing have fixed the issue.

Cause:

- After executing CLI commands and saving to file we need to send "exit" to CLI which results in another reboot
- which triggers a new connection (auto-detect)

Fix:

- Adding a timeout before starting the flash
- Blocking `auto-detect` while in backup routine as this results in another connection cycle

**Comment by nerdCopter** [comment]
i will re-test with such in mind sometime this weekend.

File: locales/en/messages.json
Comment (KarateBrot) [inline locales/en/messages.json:3550]:
```suggestion
        "message": "Flashing new firmware will wipe out all settings. We strongly recommend to save a backup before continuing.",
```

**Comment by TheIsotopes** [comment]
@haslinghuis why do we now have the same file name for `dump` and `diff`? I do not like it. 👎

**Comment by haslinghuis** [comment]
Yeah @limonspb asked to add it. We'd rather have a `dump` only imo. Also `status` is not important here.

**Comment by nerdCopter** [comment]
* tested 3 consecutive flashes successfully (without leaving tab)
* long wait after file-save :frowning_face: 
* only first flash presents the warning dialog.

**Comment by haslinghuis** [comment]
Thanks for testing @nerdCopter 

> * long wait after file-save :frowning_face: 

This should now be optimized with updated timer and added interval.

>* only first flash presents the warning dialog.

It's a daily session timer :smirk:

**Comment by nerdCopter** [comment]
thank you. i dont mind the daily-timer.  the long-wait is somewhat unnerving.  i realize it was added for OSX. Others should not have to suffer :laughing:

**Comment by haslinghuis** [comment]
Initial timeout of 500ms is definite not enough:

image

**Review by nerdCopter — APPROVED** [review]
* approving, acceptable waits for benefit of automated backups.
* tested good multiple flashes; two FC's
* Tested on Linux Debian 11

**Comment by haslinghuis** [comment]
Rebased on master

**Review by blckmn — APPROVED** [review]

**Pull request merged**
